overlord/configstate/configcore/experimental.go

```diff
@@ -0,0 +1,71 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+)
+
+func init() {
+	for _, feature := range features.KnownFeatures() {
+		supportedConfigurations["core.experimental."+feature.String()] = true
+	}
+}
+
+func validateExperimentalSettings(tr Conf) error {
+	for k := range supportedConfigurations {
+		if !strings.HasPrefix(k, "core.experimental.") {
+			continue
+		}
+		if err := validateBoolFlag(tr, strings.TrimPrefix(k, "core.")); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func handleExperimentalFlags(tr Conf) error {
+	dir := dirs.FeaturesDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return err
+	}
+	features := features.KnownFeatures()
+	content := make(map[string]*osutil.FileState, len(features))
+	for _, feature := range features {
+		if !feature.IsExported() {
```

Comment:
Ah, I see why this is public now, although I'm not sure why do we have the distinction between exported/not exported (maybe I missed something) and simply not create files for all supported featrues?

Reply:
The distinction is so that we have less conceptual problem to understand and solve. In the case of snapd reverting or upgrading, the on-disk state of exported features is constrained to this one bit (per-user mount ns). We don't have to wonder if it is ever out of sync or what happens if since the use cases are bound.

Reply:
The documentation captures this now so I think this is resolved.